### PR TITLE
Add product page

### PR DESCRIPTION
**Pull request by BryceV**

### CODE REVIEW

**Comment by BryceV** [comment]
Two things to consider UI wise with this PR:
1) The image picker has a strange inner box when you select it. The style width & height is not being overridden properly in the AddImage class (try inputting another value then 100x100). Small issue but maybe we can fix this styling issue? (resolved)
2) @ianbattin mentioned he liked all the fields (input and picker menus) the same width. Do we want a certain styling here? (resolved)
3) None of the fields have icons. We could add them if we have good icons to use.

**Comment by BryceV** [comment]
Note: There are unnecessary sections commented out like fetch/delete actions for product and a button in ArtisanDetails to get to the AddProduct page. Uncomment that button to checkout the product page.

File: src/reducers/Products.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    case 'ADD_PRODUCT':
+      return state.concat([action.product])
+    case 'GET_PRODUCTS':
+      return action.products
+    case 'DELETE_PRODUCT':
+      return state.filter(product => product.uid !== action.product.uid)
+    default:
+      return state;
+  }
+}
```
Comment (smrizzo) [inline src/reducers/Products.js:13]: Maybe change function Artisans() to function Products()

File: src/reducers/index.js [hunk 2]
```diff
@@ -5,9 +5,10 @@ import storage from 'redux-persist/lib/storage'
 import Errors from './Errors'
 import User from './User'
 import Artisans from './Artisans'
+import Products from './Products'
 
 //Then combine them all here
-const appReducer = combineReducers({ User, Errors, Artisans });
+const appReducer = combineReducers({ User, Errors, Artisans, Products });
```
Comment (abineetsingh) [inline src/reducers/index.js:11]: Rename artisan field based off the comment made in Products.js

File: src/screens/ArtisanHub/ArtisanDetail.js [hunk 3]
```diff
@@ -77,6 +77,18 @@ class ArtisanDetail extends Component {
               spinning={this.state.adding}
             />
           </View>
+          {/*
+          <View>
+            <AsyncButton
```
Comment (abineetsingh) [inline src/screens/ArtisanHub/ArtisanDetail.js:83]: This button may not be used so uncomment and test for functionality, review the scaffolded code.

File: src/screens/ArtisanHub/ArtisanDetail.js [hunk 4]
```diff
@@ -77,6 +77,18 @@ class ArtisanDetail extends Component {
...
+              title="Add Product"
+              color="green"
+              textColor="white"
+              onPress={()=>{this.props.navigation.navigate('AddProduct', {currentUID: this.props.uid})}}
+              style={styles.buttonStyle}
+              spinning={this.state.adding}
+            />
+          </View>
+          */}
```
Comment (smrizzo) [inline src/screens/ArtisanHub/ArtisanDetail.js:92]: Will need to remove the comments here

File: src/screens/Products/AddProduct.js [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+export default class AddProduct extends Component {
+  static navigationOptions = {
+    title: 'Add a Handmade Product'
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      productMainCategory: 'Handmade Apparel',
```
Comment (abineetsingh) [inline src/screens/Products/AddProduct.js]: This string is pre-filled as this is the default selection of the picker.

File: src/screens/Products/AddProduct.js [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+      productSubCategory: 'Activewear',
+
+      productMainPicture: "",
+      productTitle: "",
+      productDescription: "",
+      productGender: 'Women',
```
Comment (abineetsingh) [inline src/screens/Products/AddProduct.js]: This is a dropdown menu and the default value is Women if nothing is changed or selected.

Comment (abineetsingh) [reply src/screens/Products/AddProduct.js]: Possibly look into a default label for the picker which can be customized

File: src/screens/Products/AddProduct.js [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+      productStandardPrice: "",
+      productSellerSKU: "",
+      productQuantity: "",
+      productTime: "",
+
+      //If you adding product, assumes never sold before
+      productTimesSold: 0,
+      focusedInputs: {title: false, desc: false, price: false, sku: false, quantity: false, time: false},
```
Comment (abineetsingh) [inline src/screens/Products/AddProduct.js:29]: Orange highlighting around the box, for when a text input field is selected (focused = true), this is an interesting feature.

File: src/screens/Products/AddProduct.js [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    ImagePicker.openPicker({
+      width: 300,
+      height: 300,
+      cropping: true
+    }).then(image => {
+      this.setState({productMainPicture: image.path})
+    });
+  }
+
+  verifyFields() {
```
Comment (abineetsingh) [inline src/screens/Products/AddProduct.js:48]: A bit clunky with all the conditionals and adds complexity, possibly look into a cleaner implementation involving a boolean and what field might be missing through some type of switch case etc.

Comment (abineetsingh) [reply src/screens/Products/AddProduct.js:48]: This serves the purpose of functionality so if an alternative isn't established then just keep the code. Similar conditions are found in the add artisan.

File: src/screens/Products/AddProduct.js [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+        productionTime: this.state.productTime,
+        timesSold: this.state.productTimesSold
+      }, this.props.navigation.getParam('currentUID')).then(() => {
+        this.setState({adding: false})
+        this.props.navigation.goBack()
+      })
+    }
+  }
+
+  render() {
```
Comment (abineetsingh) [inline src/screens/Products/AddProduct.js:98]: This is a pretty massive list of categories and sub-categories, possibly shorten this or omit subcategories for simplicity. This is used in the drop down menu's so functionality is used but possibly have this in a different file etc.

File: src/screens/Products/AddProduct.js [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+    'Handmade Pet Supplies': ['Pet Apparel & Accessories', 'Pet Beds & Furniture', 'Pet Cages & Accessories', 'Pet Carriers & Travel Products', 'Pet Collars, Harnesses & Leashes', 'Pet Crates & Kennels', 'Pet Doors & Gates', 'Pet Feeding & Watering Supplies', 'Pet Flea & Tick Control', 'Pet Flood', 'Pet Grooming Supplies', 'Pet Health & Wellness', 'Pet Houses & Habitats', 'Pet Litter & Housebreaking', 'Pet Memorials & Urns', 'Pet Storage & Organization', 'Pet Toys', 'Pet Training Equipment', 'Pet Treats'], 
+    'Handmade Shoe Accessories': ['Decorative Shoe Clips', 'Insoles', 'Shoe Care Tools', 'Shoe Horns', 'Shoe Trees', 'Shoelaces', 'Shoes Bags'], 
+    'Handmade Shoes': ['Baby Booties & Crib Shoes', 'Ballet Flats', 'Boots', 'Clogs & Mules', 'First Walking Shoes', 'Loafers & Slip-Ons', 'Mary Janes', 'Oxfords', 'Pumps', 'Sandals', 'Slippers', 'Sneakers & Athletic Shoes'], 
+    'Handmade Sports & Outdoors': ['Camping & Hiking Equipment', 'Cycling Equipment', 'Fishing Equipment', 'Hunting & Shooting Equipment', 'Lawn & Playground Equipment', 'Sports & Fitness Equipment'], 
+    'Handmade Toys & Games': ['Handmade Games', 'Handmade Toys']}
+
```
Comment (smrizzo) [inline src/screens/Products/AddProduct.js]: We will need to figure out a way to store this data so have access to it. Possibly  a Json file

File: src/screens/Products/actions.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        console.log("Pushing main photo to storage")
+        
+        var st_ref = await firebase.storage()
+          .ref(`productFiles/${db_ref.key}/images/mainPicture`)
+          .putFile(data.mainPicture)
+        console.log("Finished pushing main picture to storage")
+
+        productObject.mainPictureURL = st_ref.downloadURL
+        
+        console.log("Fetching photo download url")
```
Comment (BryceV) [inline src/screens/Products/actions.js]: Change to "update url"

**Comment by BryceV** [comment]
@ianbattin @smrizzo We need to decide how the paths should be in firebase. Specifically if the products should be under the artisan id (ie products/artisanID/…)

**Review by smrizzo — COMMENTED** [review]
I just ran this branch for IOS and it appears the UI for add product did not carry over well for IOS. A few text boxes are not working, pickers are over lapping, so this will need to be fixed before we can merge this PR.

**Comment by BryceV** [comment]
> Two things to consider UI wise with this PR:
> 
> The image picker has a strange inner box when you select it. The style width & height is not being overridden properly in the AddImage class (try inputting another value then 100x100). Small issue but maybe we can fix this styling issue?
> @ianbattin mentioned he liked all the fields (input and picker menus) the same width. Do we want a certain styling here?
> None of the fields have icons. We could add them if we have good icons to use.

**Review by smrizzo — APPROVED** [review]
Overall did a good job and made some changes that we recognized today in code review.

**Pull request merged**